### PR TITLE
GCMemcard: A little cleanup.

### PR DESCRIPTION
**Pull request by AdmiralCurtiss**
Cleans up the memory card code a bit more. Tries to get rid of undefined behavior, updates naming conventions, moves code out of the header into the code file, converts enum values into enum classes or constexpr as appropriate. Should have zero functional changes.

### CODE REVIEW

**Comment by lioncash** [comment]
> As a sidenote, is some of this stuff the code is doing even legal in modern C++? 

Legal? Sure. Free of invoking undefined behavior? Absolutely not. Even in pre-modern C++.

**Comment by AdmiralCurtiss** [comment]
Yeah, I thought so. Should probably clean that up then while I'm here.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 1]
```diff
@@ -1456,17 +1456,25 @@ Header::Header(int slot, u16 size_mbits, bool shift_jis)
   calc_checksumsBE((u16*)this, 0xFE, &m_checksum, &m_checksum_inv);
 }
 
+static u32 InterpretAsU32LE(std::byte b0, std::byte b1, std::byte b2, std::byte b3)
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: This should probably go into some utility class, right?

Comment (JosJuice) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: You could probably use `swap32(const u8* data)` for this.

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Yeah I'm not sure about that. If I'm reading the standard right that's still undefined behavior if `u8` does not map to `char`, `unsigned char`, or `std::byte`. Granted, that is pretty unlikely to happen on anything but esoteric embedded hardware that can't possibly run Dolphin anyway, but y'know, if we're already rewriting it to do it right...

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: So I tried the checksum calculation with `u8` instead and yeah, that does come out a lot more readable. I dunno. I'm fine with either.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 2]
```diff
@@ -1389,6 +1387,47 @@ s32 GCMemcard::PSO_MakeSaveGameValid(const Header& cardheader, const DEntry& dir
...
+    m_serial[i] = (u8)(g_SRAM.settings_ex.flash_id[slot][i] + (u32)rand);
+    rand = (((rand * (u64)0x0000000041c64e6dULL) + (u64)0x0000000000003039ULL) >> 16);
+    rand &= (u64)0x0000000000007fffULL;
+  }
+  m_sram_bias = g_SRAM.settings.rtc_bias;
+  m_sram_language = static_cast<u32>(g_SRAM.settings.language);
+  // TODO: determine the purpose of m_unknown_2
+  // 1 works for slot A, 0 works for both slot A and slot B
+  memset(m_unknown_2.data(), 0,
+         m_unknown_2.size());  // = _viReg[55];  static vu16* const _viReg = (u16*)0xCC002000;
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: This comment has just been moved along, but do we know what it _means_?

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 3]
```diff
@@ -1389,6 +1387,47 @@ s32 GCMemcard::PSO_MakeSaveGameValid(const Header& cardheader, const DEntry& dir
...
+  memset(this, 0xFF, BLOCK_SIZE);
+  m_size_mb = size_mbits;
+  m_encoding = shift_jis ? 1 : 0;
+  u64 rand = Common::Timer::GetLocalTimeSinceJan1970() - ExpansionInterface::CEXIIPL::GC_EPOCH;
+  m_format_time = rand;
+  for (int i = 0; i < 12; i++)
+  {
+    rand = (((rand * (u64)0x0000000041c64e6dULL) + (u64)0x0000000000003039ULL) >> 16);
+    m_serial[i] = (u8)(g_SRAM.settings_ex.flash_id[slot][i] + (u32)rand);
+    rand = (((rand * (u64)0x0000000041c64e6dULL) + (u64)0x0000000000003039ULL) >> 16);
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Any point in making those `constexpr`? Not sure what name to give them tho.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 4]
```diff
@@ -1494,6 +1503,22 @@ std::string DEntry::GCI_FileName() const
...
+  auto [checksum_sum, checksum_inv] = CalculateChecksums();
+  m_checksum = checksum_sum;
+  m_checksum_inv = checksum_inv;
+}
+
+std::pair<u16, u16> Header::CalculateChecksums() const
+{
+  static_assert(std::is_trivially_copyable<Header>::value);
+
+  std::array<u8, 0x1FC> raw;
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Thoughts on using a `constexpr` based on `Header` for this size instead? Something like `offsetof(Header, m_update_counter) - offsetof(Header, m_serial)` to make it clear what parts will be checksummed.
And if we decide to go with this, the same for the others.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 2 above)
Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Not sure? I guess it's meant to imply that it copies the value from the `VI_DTV_STATUS` register, but I don't think that makes a lot of sense.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 4 above)
Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Yeah, that makes sense.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 5]
```diff
@@ -595,57 +593,100 @@ std::optional<DEntry> GCMemcard::GetDEntry(u8 index) const
...
+  m_map[current - 5] = 0xFFFF;
+  m_last_allocated_block = current;
+  m_free_blocks = m_free_blocks - length;
+  FixChecksums();
+  return starting;
+}
+
+std::pair<u16, u16> BlockAlloc::CalculateChecksums() const
+{
+  static_assert(std::is_trivially_copyable<BlockAlloc>::value);
```
Comment (lioncash) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]:
```suggestion
  static_assert(std::is_trivially_copyable<BlockAlloc>());
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 1 above)
Comment (lioncash) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: > Yeah I'm not sure about that. If I'm reading the standard right that's still undefined behavior if u8 does not map to `char`, `unsigned char`, or `std::byte`

Dolphin would not run at all on platforms where this isn't the case (just as we wouldn't run on platforms where `float` and `double` aren't 32-bit and 64-bit respectively)

**Comment by AdmiralCurtiss** [comment]
By the by, have we ever considered Dolphin support on big endian platforms? A decent amount of code in this class assumes native little endian.

**Comment by AdmiralCurtiss** [comment]
Should we also modify the constructors to use standard initialization instead of memset? I started doing that but stopped myself after noticing that a one-liner `memset(this, 0xFF, DENTRY_SIZE)` transformed into this thing:

~~~
DEntry::DEntry()
    : m_unused_1(0xFFu), m_banner_and_icon_flags(0xFFu), m_modification_time(0xFFFFFFFFu),
      m_image_offset(0xFFFFFFFFu), m_icon_format(0xFFFFu), m_animation_speed(0xFFFFu),
      m_file_permissions(0xFFu), m_copy_counter(0xFFu), m_first_block(0xFFFFu),
      m_block_count(0xFFFFu), m_comments_address(0xFFFFFFFFu)
{
  m_gamecode.fill(0xFF);
  m_makercode.fill(0xFF);
  m_filename.fill(0xFF);
  m_unused_2.fill(0xFF);
}
~~~

And as far as I can tell memset is totally fine for trivially copyable types.

**Comment by BhaaLseN** [comment]
It probably becomes a lot less verbose if you just in-class initialize them:
```c++
struct DEntry
{
  DEntry();
  std::string GCI_FileName() const;
  static constexpr std::array<u8, 4> UNINITIALIZED_GAMECODE{{0xFF, 0xFF, 0xFF, 0xFF}};

  std::array<u8, 4> m_gamecode{{ 0xFF, 0xFF, 0xFF, 0xFF }};
  std::array<u8, 2> m_makercode{{ 0xFF, 0xFF }};
  u8 m_unused_1 = 0xFF;
  u8 m_banner_and_icon_flags = 0xFF;
  // etc...
};
```

Kinda +/-0 on that though...

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 1 above)
Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Okay, removed the whole thing and just using u8.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 4 above)
Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Done.

**Comment by AdmiralCurtiss** [comment]
Regarding commit eea5a306c30d35fa5d84b27452d86f291dd1fd0a, I find it a bit weird how the code goes through great pain to return a somewhat descriptive return code but all the calling code ever only checks for `SUCCESS`. Should we simplify the code here to just return a true/false, or should we give better error messages on the outside?

**Comment by JMC47** [comment]
Any update on this?  I don't see any approvals on it so I don't want to merge it, but, our GC memcard code probably needs the cleanups based on what I know of it.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 6]
```diff
@@ -298,25 +301,25 @@ u32 GCMemcard::TestChecksums() const
 
   u32 results = 0;
 
-  calc_checksumsBE((u16*)&m_header_block, 0xFE, &csum, &csum_inv);
+  std::tie(csum, csum_inv) = m_header_block.CalculateChecksums();
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Now that we're officially C++17, you could get back to using structured bindings ;)

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Haha, sure, why not.

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Alright, done.

**Comment by AdmiralCurtiss** [comment]
From my side I think this is good to go, there's still sketchy stuff in here but nothing that wasn't already sketchy before, and I think adding more on top of this would just make it harder to review. Let's refactor this step by step.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 7]
```diff
@@ -1457,17 +1455,25 @@ Header::Header(int slot, u16 size_mbits, bool shift_jis)
...
+
+  u32 serial1 = 0;
+  u32 serial2 = 0;
+  for (size_t i = 0; i < raw.size(); i += 8)
   {
-    memcpy(&serial[i], (u8*)this + (i * 4), 4);
+    u32 tmp1, tmp2;
+    memcpy(&tmp1, &raw[i + 0], sizeof(u32));
+    memcpy(&tmp2, &raw[i + 4], sizeof(u32));
+    serial1 ^= tmp1;
```
Comment (leoetlino) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]:
```cpp
serial1 ^= Common::BitCast<u32>(&raw[i + 0]);
```

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: I assume you mean `BitCastPtr`?

Comment (leoetlino) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Whoops, yeah, sorry for the confusion.

**Pull request merged**